physical/postgresql.go

```diff
@@ -0,0 +1,194 @@
+        IF found THEN
+            RETURN;
+        END IF;
+        -- not there, so try to insert the key
+        -- if someone else inserts the same key concurrently,
+        -- we could get a unique-key failure
+        BEGIN
+            INSERT INTO ` + table + `(vault_key, vault_value) VALUES (key, value);
+            RETURN;
+        EXCEPTION WHEN unique_violation THEN
+            -- Do nothing, and loop to try the UPDATE again.
+        END;
+    END LOOP;
+END;
+$$
+LANGUAGE plpgsql;`
+
+		if _, err := db.Exec(create_upsert); err != nil {
+			return nil, fmt.Errorf("failed to create upsert function: %v", err)
+		}
+
+		put_statement = "SELECT upsert_vault($1, $2)"
+	}
+
+	// Setup the backend.
+	m := &PostgreSQLBackend{
+		table:      table,
+		client:     db,
+		statements: make(map[string]*sql.Stmt),
+	}
+
+	// Prepare all the statements required
+	statements := map[string]string{
+		"put":    put_statement,
+		"get":    "SELECT vault_value FROM " + table + " WHERE vault_key = $1",
+		"delete": "DELETE FROM " + table + " WHERE vault_key = $1",
+		"list":   "SELECT vault_key FROM " + table + " WHERE vault_key LIKE $1",
+	}
+	for name, query := range statements {
+		if err := m.prepare(name, query); err != nil {
+			return nil, err
+		}
+	}
+	return m, nil
+}
+
+// prepare is a helper to prepare a query for future execution
+func (m *PostgreSQLBackend) prepare(name, query string) error {
+	stmt, err := m.client.Prepare(query)
+	if err != nil {
+		return fmt.Errorf("failed to prepare '%s': %v", name, err)
+	}
+	m.statements[name] = stmt
+	return nil
+}
+
+// Put is used to insert or update an entry.
+func (m *PostgreSQLBackend) Put(entry *Entry) error {
+	defer metrics.MeasureSince([]string{"postgres", "put"}, time.Now())
+
+	_, err := m.statements["put"].Exec(entry.Key, entry.Value)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Get is used to fetch and entry.
+func (m *PostgreSQLBackend) Get(key string) (*Entry, error) {
+	defer metrics.MeasureSince([]string{"postgres", "get"}, time.Now())
+
+	var result []byte
+	err := m.statements["get"].QueryRow(key).Scan(&result)
+	if err == sql.ErrNoRows {
+		return nil, nil
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	ent := &Entry{
+		Key:   key,
+		Value: result,
+	}
+	return ent, nil
+}
+
+// Delete is used to permanently delete an entry
+func (m *PostgreSQLBackend) Delete(key string) error {
+	defer metrics.MeasureSince([]string{"postgres", "delete"}, time.Now())
+
+	_, err := m.statements["delete"].Exec(key)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// List is used to list all the keys under a given
+// prefix, up to the next prefix.
+func (m *PostgreSQLBackend) List(prefix string) ([]string, error) {
+	defer metrics.MeasureSince([]string{"postgres", "list"}, time.Now())
+
+	// Add the % wildcard to the prefix to do the prefix search
+	likePrefix := prefix + "%"
+	rows, err := m.statements["list"].Query(likePrefix)
```

Comment:
Probably worth calling `defer rows.Close()` here in case you bail out of the loop early. (https://golang.org/pkg/database/sql/#Rows.Close)